Support to use `tsconfig-paths` in TypeScript wdio config file

**Environment (please complete the following information):**
 - **WebdriverIO version:** [7.0.7]
 - **Mode:** [WDIO Testrunner]
 - **If WDIO Testrunner, running sync/async:** [sync]
 - **Node.js version:** [14.15.5]
 - **NPM version:** [yarn 1.22.10]
 - **Platform name and version:** [macOS Mojave]

**Config of WebdriverIO**
filename: wdio.conf.ts

```
import module1 from 'modules/module/module1'

const config: WebdriverIO.Config = {
    runner: 'local',
    specs: [
        './specs/**/*.ts'
    ],
    capabilities: [{
        browserName: 'chrome',
    }],
    framework: 'mocha',
    reporters: ['spec'],
    mochaOpts: {
        ui: 'bdd',
    },
    autoCompileOpts: {
        autoCompile: true,
        tsNodeOpts: {
            transpileOnly: true,
            project: 'tsconfig.json',
            require: ['tsconfig-paths'] // tried also ['tsconfig-paths/register']
        }
    },
}
export {config}
```

**Describe the bug**
When I try to import some module described in paths in tsconfig.json in scope of WDIO TypeScript config file using 'tsconfig-paths' I'm getting Error: Cannot find module 'modules/module/module1'

**To Reproduce**
Have following in tsconfig.json
```
    "baseUrl": "./",
    "paths": {
      "modules/*": ["modules/*"],
    }
```

Some module in specified path.

Run `wdio run wdio.conf.ts`

**Expected behavior**
WDIO will find module, like it's working now with following solution:

file: wdio.conf.js

```
require('ts-node').register({ transpileOnly: true })
require('tsconfig-paths').register()
module.exports = require('./wdio.conf.ts')
```

file: wdio.conf.ts
```
import module1 from 'modules/module/module1'

const config: WebdriverIO.Config = {
    runner: 'local',
    specs: [
        './specs/**/*.ts'
    ],
    capabilities: [{
        browserName: 'chrome',
    }],
    framework: 'mocha',
    reporters: ['spec'],
    mochaOpts: {
        ui: 'bdd',
    },
}
export {config}
```

Run `wdio run wdio.conf.js`

**Log**

Error: Cannot find module 'modules/module/module1'
Require stack:
- /Users/jkostrhun/wdio.conf.ts
- /Users/jkostrhun/node_modules/@wdio/config/build/lib/FileSystemPathService.js
- /Users/jkostrhun/node_modules/@wdio/config/build/lib/ConfigParser.js
- /Users/jkostrhun/node_modules/@wdio/config/build/index.js
- /Users/jkostrhun/node_modules/@wdio/cli/build/launcher.js
- /Users/jkostrhun/node_modules/@wdio/cli/build/index.js
- /Users/jkostrhun/node_modules/@wdio/cli/bin/wdio.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/Users/jkostrhun/wdio.conf.ts:1:1)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Module.m._compile (/Users/jkostrhun/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/jkostrhun/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (internal/modules/cjs/loader.js:928:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/jkostrhun/wdio.conf.ts',
    '/Users/jkostrhun/node_modules/@wdio/config/build/lib/FileSystemPathService.js',
    '/Users/jkostrhun/node_modules/@wdio/config/build/lib/ConfigParser.js',
    '/Users/jkostrhun/node_modules/@wdio/config/build/index.js',
    '/Users/jkostrhun/node_modules/@wdio/cli/build/launcher.js',
    '/Users/jkostrhun/node_modules/@wdio/cli/build/index.js',
    '/Users/jkostrhun/node_modules/@wdio/cli/bin/wdio.js'
  ]
}

--- a/examples/wdio.conf.js
+++ b/examples/wdio.conf.js
@@ -193,6 +193,12 @@ exports.config = {
             project: 'tsconfig.json'
         },
 
+        // If you have tsconfig-paths installed and provide a tsConfigPathsOpts
+        // option, it will be automatically registered during bootstrap.
+        tsConfigPathsOpts: {
+            baseUrl: './'
+        },
+
         // Configure how @babel/register is automatically included when present (and ts-node isn't)
         babelOpts: {}
     },

--- a/packages/wdio-config/package.json
+++ b/packages/wdio-config/package.json
@@ -32,6 +32,7 @@
     "access": "public"
   },
   "devDependencies": {
-    "minimatch": "^3.0.4"
+    "minimatch": "^3.0.4",
+    "tsconfig-paths": "^3.9.0"
   }
 }

--- a/packages/wdio-config/src/utils.ts
+++ b/packages/wdio-config/src/utils.ts
@@ -96,18 +96,39 @@ export interface ModuleRequireService {
 }
 
 export function loadAutoCompilers(autoCompileConfig: Options.AutoCompileConfig, requireService: ModuleRequireService) {
-    return autoCompileConfig.autoCompile && (
-        loadTypeScriptCompiler(autoCompileConfig.tsNodeOpts, requireService)
-        ||
-        loadBabelCompiler(autoCompileConfig.babelOpts, requireService)
+    return (
+        autoCompileConfig.autoCompile &&
+        (
+            loadTypeScriptCompiler(
+                autoCompileConfig.tsNodeOpts,
+                autoCompileConfig.tsConfigPathsOpts,
+                requireService
+            )
+            ||
+            loadBabelCompiler(
+                autoCompileConfig.babelOpts,
+                requireService
+            )
+        )
     )
 }
 
-export function loadTypeScriptCompiler (tsNodeOpts: RegisterOptions = {}, requireService: ModuleRequireService) {
+export function loadTypeScriptCompiler (
+    tsNodeOpts: RegisterOptions = {},
+    tsConfigPathsOpts: Options.TSConfigPathsOptions | undefined,
+    requireService: ModuleRequireService
+) {
     try {
         requireService.resolve('ts-node') as any
         (requireService.require('ts-node') as any).register(tsNodeOpts)
         log.debug('Found \'ts-node\' package, auto-compiling TypeScript files')
+
+        if (tsConfigPathsOpts) {
+            log.debug('Found \'tsconfig-paths\' options, register paths')
+            const tsConfigPaths = require('tsconfig-paths')
+            tsConfigPaths.register(tsConfigPathsOpts)
+        }
+
         return true
     } catch (e) {
         return false

--- a/packages/wdio-config/tests/configparser.test.ts
+++ b/packages/wdio-config/tests/configparser.test.ts
@@ -1,5 +1,6 @@
 import path from 'path'
 import tsNode from 'ts-node'
+import tsConfigPath from 'tsconfig-paths'
 import logger from '@wdio/logger'
 import ConfigParser from '../src/lib/ConfigParser'
 import MockFileContentBuilder, { MockFileContent } from './lib/MockFileContentBuilder'
@@ -19,6 +20,10 @@ const FIXTURES_CUCUMBER_FEATURE_B_LINE_7 = path.resolve(FIXTURES_PATH, 'test-b.f
 const INDEX_PATH = path.resolve(__dirname, '..', 'src', 'index.ts')
 
 jest.mock('ts-node', () => ({
+    register: jest.fn()
+}))
+
+jest.mock('tsconfig-paths', () => ({
     register: jest.fn()
 }))
 
@@ -165,6 +170,7 @@ describe('ConfigParser', () => {
             beforeEach(() => {
                 (log.debug as jest.Mock).mockClear()
                 ;(tsNode.register as jest.Mock).mockClear()
+                ;(tsConfigPath.register as jest.Mock).mockClear()
                 process.env.THROW_BABEL_REGISTER = '1'
             })
 
@@ -266,9 +272,37 @@ describe('ConfigParser', () => {
                     .withTsNodeModule(tsNodeRegister).build()
                 configParser.addConfigFile('tests/cool.conf')
                 configParser.autoCompile()
+                expect(tsConfigPath.register).toBeCalledTimes(0)
                 expect(tsNodeRegister).toBeCalledTimes(1)
                 expect(tsNodeRegister).toHaveBeenCalledWith( {
                     'transpileOnly': false
+                } )
+            })
+
+            it('bootstraps tsconfig-paths if options are given', function () {
+                let configFileContents = MockFileContentBuilder.FromRealConfigFile(FIXTURES_CONF_RDC).withTheseContentsMergedOn(
+                    {
+                        config: {
+                            autoCompileOpts: {
+                                tsConfigPathsOpts: {
+                                    base: '/foo/bar'
+                                }
+                            }
+                        }
+                    }
+                ).build()
+                const tsNodeRegister = jest.fn()
+                const configParser = ConfigParserBuilder
+                    .withBaseDir(path.join(__dirname, '/tests/'))
+                    .withFiles([
+                        ...MockedFileSystem_OnlyLoadingConfig(path.join(__dirname, '/tests/'), '/path/to/config'),
+                        FileNamed(path.join(__dirname, '/tests/tests/cool.conf')).withContents(JSON.stringify(configFileContents))
+                    ])
+                    .withTsNodeModule(tsNodeRegister).build()
+                configParser.addConfigFile('tests/cool.conf')
+                configParser.autoCompile()
+                expect(tsConfigPath.register).toHaveBeenCalledWith( {
+                    base: '/foo/bar'
                 } )
             })
 

--- a/packages/wdio-types/src/Options.ts
+++ b/packages/wdio-types/src/Options.ts
@@ -410,10 +410,18 @@ export interface Testrunner extends Hooks, Omit<WebdriverIO, 'capabilities'>, We
     autoCompileOpts?: AutoCompileConfig
 }
 
+export interface TSConfigPathsOptions {
+    baseUrl: string
+    paths: Record<string, string[]>
+    mainFields?: string[]
+    addMatchAll?: boolean
+}
+
 export interface AutoCompileConfig {
     autoCompile?: boolean
     tsNodeOpts?: RegisterOptions
     babelOpts?: Record<string, any>
+    tsConfigPathsOpts?: TSConfigPathsOptions
 }
 
 export interface MultiRemote extends Omit<Testrunner, 'capabilities'> {

--- a/website/docs/ConfigurationFile.md
+++ b/website/docs/ConfigurationFile.md
@@ -268,6 +268,11 @@ exports.config = {
             transpileOnly: true,
             project: 'tsconfig.json'
         },
+        // If you have tsconfig-paths installed and provide a tsConfigPathsOpts
+        // option, it will be automatically registered during bootstrap.
+        tsConfigPathsOpts: {
+            baseUrl: './'
+        },
         //
         // If @babel/register is installed, you can customize how options are passed to it here:
         // Any valid @babel/register config option is allowed.

--- a/website/docs/TypeScript.md
+++ b/website/docs/TypeScript.md
@@ -11,7 +11,31 @@ You will need [`typescript`](https://github.com/microsoft/TypeScript) and [`ts-n
 $ npm install typescript ts-node --save-dev
 ```
 
-The minimum TypeScript version is 3.8.3.
+The minimum TypeScript version is `v4.0.5`.
+
+## Configuration
+
+You can provide custom `ts-node` and `tsconfig-paths` options through your `wdio.conf.ts`, e.g.:
+
+```ts title="wdio.conf.ts"
+export const config = {
+    // ...
+    autoCompileOpts: {
+        autoCompile: true,
+        // see https://github.com/TypeStrong/ts-node#cli-and-programmatic-options
+        // for all available options
+        tsNodeOpts: {
+            transpileOnly: true,
+            project: 'tsconfig.json'
+        },
+        // tsconfig-paths is only used if "tsConfigPathsOpts" are provided, if you
+        // do please make sure "tsconfig-paths" is installed as dependency
+        tsConfigPathsOpts: {
+            baseUrl: './'
+        }
+    }
+}
+```
 
 ## Framework Setup
 
@@ -49,7 +73,7 @@ And your `tsconfig.json` needs the following:
 </Tabs>
 
 Please avoid importing `webdriverio` or `@wdio/sync` explicitly.
-`WebdriverIO` and `WebDriver` types are accessible from anywhere once added to `types` in `tsconfig.json`.
+`WebdriverIO` and `WebDriver` types are accessible from anywhere once added to `types` in `tsconfig.json`. If you use additional WebdriverIO services or plugins, please also add them to the `types` list as many provide additional typings.
 
 ## Framework types
 

